R/pkg/R/mllib_recommendation.R

```diff
@@ -82,6 +82,10 @@ setClass("ALSModel", representation(jobj = "jobj"))
 #' statsS <- summary(modelS)
 #' }
 #' @note spark.als since 2.1.0
+#' @note the input rating dataframe to the ALS implementation should not be indeterminate.
```

Comment:
I still think we need to say "nondeterministic" and give an example (randomSplit), but also tell people how to fix it. Define a partitioning? sort order?

Reply:
A checkpoint or a sort before sampling can help. Sampled RDD is nondeterministic when its input RDD is unordered.

Reply:
Is there any way to make, say, the result of randomSplit deterministic after it has been computed? I understand the usual answer of caching isn't 100% foolproof. But if there is no way to do it, hm, is it worth warning? because this is just generally true of lots of things in Spark

Reply:
If randomSplit or sample is computed, the only way to make it deterministic is to checkpoint. But obviously we can't checkpoint for users.

I think it is good to leave a clue so users can know what is going on when hit the ArrayIndexOutOfBoundsException during fitting ALS model.

Since we don't want to break existing user code, a warning is the least thing we can do?

I am also like to catch the exception and re-throw a meaningful message to users.

Reply:
But what if you lose the checkpoint? isn't this the same issue? At some level the answer is "you can't really fix this, anywhere", right? in practice, the fairly well understood caching/checkpoint mechanism works, everywhere. It seems inconsistent to just address this for ALS, as if it's not the same issue everywhere. It also seems hard to warn without providing any pointer to the solution, if there is one, but I can see that a warning is better than nothing.

If you're trying to fix a specific problem, maybe indeed detect the problem in question (a specific AIOOBE) and rewrap it, sure.

Reply:
I think checkpoint is relatively reliable. In case of checkpoint loss, Spark job fails without rerun. So you should not get an inconsistent data once you do checkpoint.

We have two ways to fix it, one is checkpoint, another is to sort data before sample/randomSplit. I added into the updated note.

Sounds like targeting a specific problem here is better. I do the catching AIOOBE thing and remove the warning as it seems not too much useful.